lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,303 @@
+			'internaltype'=> 'string',
+			'default'=> '',
+		//	'public' => false ,
+ 			'required'=> false
+		),
+	);
+
+	public function getConfigBE() : array
+	{
+		$list = [];
+
+		foreach( $this->beConfig as $key => $config ) {
+			$list[$key] = new \Aimeos\MW\Criteria\Attribute\Standard( $config );
+		}
+
+		return $list;
+	}
+
+	protected function getProvider() : \Omnipay\Common\GatewayInterface
+	{
+		if( !isset( $this->provider ) )
+		{
+			$this->provider = OPay::create( 'PayPal_Rest' );
+			$this->provider->setTestMode( (bool) $this->getValue( 'testmode', false ) );
+			$this->provider->initialize( $this->getServiceItem()->getConfig() );
+		}
+
+		return $this->provider;
+	}
+
+
+	protected function sendRequest( \Aimeos\MShop\Order\Item\Iface $order, array $data ) : \Omnipay\Common\Message\ResponseInterface
+	{
+		$provider = $this->getProvider();
+
+		if( $this->getValue( 'authorize', false ) && $provider->supportsAuthorize() )
+		{
+			$response = $provider->authorize( $data )->send();
+			$order->setPaymentStatus( Status::PAY_AUTHORIZED );
+		}
+		else
+		{
+			$response = $provider->purchase( $data )->send();
+			$order->setPaymentStatus( Status::PAY_RECEIVED );
+		}
+
+		return $response;
+	}
+
+	protected function getPaymentForm( \Aimeos\MShop\Order\Item\Iface $order, array $params ) : \Aimeos\MShop\Common\Helper\Form\Iface
+	{
+		$list = [];
+		$feConfig = $this->feConfig;
+		$baseItem = $this->getOrderBase( $order->getBaseId(), \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS );
+		$addresses = $baseItem->getAddress( \Aimeos\MShop\Order\Item\Base\Address\Base::TYPE_PAYMENT );
+
+		$year = date( 'Y' );
+
+		foreach( $feConfig as $key => $config ) {
+			$list[$key] = new \Aimeos\MW\Criteria\Attribute\Standard( $config );
+		}
+
+		// /* section of preparation to call paypal plus form
+		$parts = \Aimeos\MShop\Order\Item\Base\Base::PARTS_SERVICE
+		| \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+		| \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS;
+
+		$base = $this->getOrderBase( $order->getBaseId(), $parts );
+		$data = $this->getData( $base, $order->getId(), $params );
+		$urls = $this->getPaymentUrls();
+
+		/* 
+		$this->getPaymentUrls() returns cancelUrl same as in returnUrl 
+		which is wrong ! so we added this line and comment it till we get sure about it !
+		// $data['cancelUrl'] = $this->getConfigValue( 'cancelUrl', '' ) ;
+
+		if(empty($data['cancelUrl'])) {
+			$data['cancelUrl'] = $_SERVER['HTTP_ORIGIN'] ;
+		}
+		*/
+
+		try
+		{
+			$response = $this->sendRequest( $order, $data );
+			$approvalUrl = "" ;
+			$testmode = "" ;
+			$countryid = "" ;
+
+			if( $response->isSuccessful() )
+			{
+				$this->setOrderData( $order, ['Transaction' => $response->getTransactionReference()] );
+				$this->saveRepayData( $response, $base->getCustomerId() );
+
+				$status = $this->getValue( 'authorize', false ) ? Status::PAY_AUTHORIZED : Status::PAY_RECEIVED;
+				$this->saveOrder( $order->setPaymentStatus( $status ) );
```

Comment:
This seems to be wrong here because the PayPalPlus form hasn't been rendered yet but you already set the order status to "authorized" or "received".

Reply:
yes it seems wrong , but it working fine .. and I will explain that :
you are right that the status set as "received" before rendering of PPP , but that will be corrected by updateSync() which change the status to "unfinished" if the buyer cancel the PPP process , or keep it "received" if he complete the purchase successfully .. 
of course that updateSync() had worked just -as I think and not sure yet-  because of PPP returned urls like "cancelUrl and acceptedUrl ..etc" and not by normal execution workflow that other payment methods do .. (this point needs more work to be clear and optimized by me )
I have finished all other notes , but I didn't push the changes because I am still working on this point ..
so should I keep working on it , or return back to the last submitted code ?

Reply:
It's wrong nevertheless. If you choose PPP option, go to the process page an close the browser, the order is paid and you get the products for free.

Reply:
you are right , and sorry for delay .. 

I was working in this point and thanks God it is solved right now and everything are working as expected .. 

sure firstly I set the state of order is "unfinished" ..

the problem was that I couldn't get the result of outSite PPP JS process ...
solution : I have used $response of  $provider->completeAuthorize($params) .. 
BTY : I dissevered that our payerID is not accepted from PayPal !  ... and the conflict was because I missed that PayPal uses two phases -not one phase like other payment gateways!- process to complete the PPP payment ..
first to get approvalUrl , payerID and token ..  and this payerID should I use in second phase ! so I corrected them and no need for override getData() any more !
I have set payerID like so : $params['PayerID'] = $request->getQueryParams()['PayerID'] ; in updateSybc() 

then, everything are running as expected except two points : 
1. old checking ($response->getTransactionId() != $order->getId() )  not working ! as $response->getTransactionId() return null always ! and the solution is to get TransactionId from the request parameters as following : $response->getRequest()->getTransactionId()
2. method_exists($response, 'isCancelled') always null as it doesn't consider what happened after go out of Aimeos , so it consider the status of cancel during PPP process as "Refused" ! .. and I think it is wrong so I have add this concatenation like so : || ( !empty($response->getData()['name']) && $response->getData()['name'] == "PAYMENT_NOT_APPROVED_FOR_EXECUTION") and now it cancel during PPP is considered like "cancel" state ..

last thing , I have set PSR12 rules and corrected coding style I will correct anything in future after get your opinion of what I have mentioned previously 

thanks in advance ..

Reply:
The missing transaction ID is a weakness in Paypal_Rest driver and you should create a PR there to set the transaction ID from the request to the response. In the meantime, we can overwrite our updateSync() method and use `$response->getRequest()->getTransactionId() === $order->getId()`

Reply:
Same for `isCancelled()`. You should create a PR in the Omnipay Paypal_Rest driver that fixes that in the driver to return true for isCancelled() if PAYMENT_NOT_APPROVED_FOR_EXECUTION is returned by PayPal

Reply:
ok , I will do that